src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -1027,14 +1027,21 @@ public IQueryCache GetQueryCache(string cacheRegion)
 				return null;
 			}
 
-			return queryCaches.GetOrAdd(
+			var wasAdded = false;
+			// The factory may be run concurrently by threads trying to get the same region.
+			// But the GetOrAdd will yield the same cache for all threads, so in case of add, use
+			// it to update allCacheRegions
+			var cache = queryCaches.GetOrAdd(
 				cacheRegion,
 				cr =>
 				{
-					IQueryCache currentQueryCache = settings.QueryCacheFactory.GetQueryCache(cacheRegion, updateTimestampsCache, settings, properties);
-					allCacheRegions[currentQueryCache.RegionName] = currentQueryCache.Cache;
+					var currentQueryCache = settings.QueryCacheFactory.GetQueryCache(cacheRegion, updateTimestampsCache, settings, properties);
+					wasAdded = true;
 					return currentQueryCache;
 				});
+			if (wasAdded)
+				allCacheRegions[cache.RegionName] = cache.Cache;
```

Comment:
This fix does not avoid building many caches for the same region, but it ensures we store the same everywhere.
We should use https://stackoverflow.com/a/31637510/1178314 if we want to build only one cache per region. This will introduce additional locks, so it would have a performance impact.

Reply:
I now realize there is another bug here (already there without that PR): if the user asks for a query cache in a region used for an entity cache, a new cache will be built and will "replace" the entity cache in the global cache dictionary.

(Some cache provider may yield the same cache, having their own "cache of cache by region", especially for configured region.)

I see many ways of fixing that other trouble:
 * Adding some prefix to region name before storing in the global cache. Brittle, may breaks expectations on how to retrieve a query cache through `GetSecondLevelCacheRegion`.
 * In the `GetOrAdd` value factory, try get the cache from the global cache and pass it for reuse to `IQueryCacheFactory`. Require a new method on this interface and implies obsoleting the previous one.
 * In the `GetOrAdd` value factory, try get the cache from the global cache and throw if already there, treating this as an error and asking the user for not re-using an entity region as a query region. This will require ensuring the value factory is executed only once per query cache with the `Lazy<T>` solution by example. This will be a breaking change for user already using the same region for query caches and entity caches.
 * Leave this trouble in place. That global cache dictionary has no significant usages currently. (At least no one seems to have ever complained.)

Reply:
Looking more into cache initialization on factory build, there are even more issues, such as not respecting the cache concurrency if many entities use the same region, but with different concurrences. The first to be configured will set the concurrency, and the others will reuse it silently. But for collections, the trouble will not occurs, each will have their expected concurrency even if sharing the same region.

I will likely do another PR for cleaning the cache initialization logic. On Hibernate side, things have quite changed since 5.2 about that subject. They have moved most of the cache handling out of the session factory. Porting this should likely be done only for 6.0.